src/eval.ml

```diff
@@ -165,6 +169,91 @@ and eq_modulo : term -> term -> bool = fun a b ->
+and tree_walk : Dtree.t -> stack -> (term * stack) option = fun itree istk ->
+  (* Count maximum number of stored items to initialize variable array *)
+  let capacity = Dtree.iter ~do_leaf:(fun _ _ -> 0)
+    ~do_node:(fun _ store subr abstr defr ->
+      let maxch = if subr = [] then 0
+        else List.extremum (>) (snd (List.split subr)) in
+      let maxchd = match defr with
+        | None     -> maxch
+        | Some(dr) -> max dr maxch in
+      let maxchdab = match abstr with
+        | None        -> maxchd
+        | Some(_, ar) -> max ar maxchd in
+      if store then succ maxchdab else maxchdab)
+    ~fail:0 itree in
+  let vars = Array.init capacity (fun _ -> Patt(None, "", [| |])) in
+
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array . *)
+  let rec walk : Dtree.t -> stack -> int -> (term * stack) option =
+    fun tree stk cursor ->
+      match tree with
+      | Fail                                  -> None
+      | Leaf(env_builder, act)                ->
+         assert (IntMap.cardinal env_builder = Bindlib.mbinder_arity act) ;
+        (* [pre_env] is the same as [env] but without binders *)
+        let pre_env = IntMap.fold (fun pos sl pe ->
+          IntMap.add sl vars.(pos) pe) env_builder IntMap.empty in
+        let env = Array.init (IntMap.cardinal pre_env) (fun _ -> TE_None) in
+        IntMap.iter (fun sl te ->
+          let inject _ = te in
+          let b = Bindlib.raw_mbinder [| |] [| |] 0 mkfree inject in
+          env.(sl) <- TE_Some(b)) pre_env ;
+        Some(Bindlib.msubst act env, stk)
+      | Node({ swap ; children ; store ; default ; abstspec }) ->
+         if stk = [] then None else (* If stack too short, quit *)
+         (* Pick the right term in the stack *)
+         let stk = match swap with
+           | None    -> stk
+           | Some(i) -> List.bring i stk in
+         let examined = List.hd stk in
+         if not (fst Pervasives.(!examined))
+         then Pervasives.(examined := (true, unfold (snd !examined))) ;
+         (* ^ Ought to be lightened, as well as [stack] data structure *)
+         (* Remove [Appl] nodes *)
+         let stk = match snd Pervasives.(!examined) with
+           | Appl(_, _) as a ->
+              let t, la = Basics.get_args a in
+              let unfolded = List.map (fun e -> Pervasives.ref (false, e))
+                la in
+              (Pervasives.ref (false, unfold t)) :: unfolded @ (List.tl stk)
+           | _               -> stk in
+         (* Store hd of stack if needed *)
+         if store then vars.(cursor) <- (snd Pervasives.(!(List.hd stk))) ;
+         let ncurs = if store then succ cursor else cursor in
+         (* Fetch the right subtree and the new stack *)
+         let matched, tlstk = match snd Pervasives.(!(List.hd stk)) with
+           | Symb(_, _) as s ->
+              let matched_on_cons = List.assoc_eq Basics.eq s children in
```

Comment:
This is not very efficient to use a (term*tree)list structure for children here. Why not using at least a tree StrMap.t?